File: external/cec2006_src/cec2006-benchmarks-master/code/python/g21.py
import numpy as np

def g21_objective(x):
    """
    CEC 2006 g21 constrained optimization problem objective function.
    
    Parameters:
    x (array-like): Decision variables [x1, x2, ..., x7]
    
    Returns:
    float: Objective function value
    """
    return np.prod(x)

def g21_constraints(x):
    """
    CEC 2006 g21 constraints.
    
    Parameters:
    x (array-like): Decision variables [x1, x2, ..., x7]
    
    Returns:
    list: Constraint values (should be <= 0 for feasible)
    """
    g = []
    g.append(sum(x) - 1)
    return g

def g21_bounds():
    """
    Returns the bounds for g21 problem.
    
    Returns:
    list: [(lower, upper), ...] for each variable
    """
    return [(0, 10)] * 7

# Example usage
if __name__ == "__main__":
    # Example solution (best known)
    x_opt = [0.1] * 7  # Approximate
    f_opt = g21_objective(x_opt)
    g_opt = g21_constraints(x_opt)
    
    print(f"Objective value: {f_opt}")
    print(f"Constraints: {g_opt}")
    print(f"All constraints satisfied: {all(g <= 0 for g in g_opt)}")

File: external/cec2006_src/cec2006-benchmarks-master/code/matlab/g21.m
function f = g21_objective(x)
% CEC 2006 g21 constrained optimization problem objective function.
%
% Parameters:
% x: Decision variables [x1, x2, ..., x7]
%
% Returns:
% f: Objective function value

f = prod(x);

function g = g21_constraints(x)
% CEC 2006 g21 constraints.
%
% Parameters:
% x: Decision variables [x1, x2, ..., x7]
%
% Returns:
% g: Constraint values (should be <= 0 for feasible)

g = sum(x) - 1;

function bounds = g21_bounds()
% Returns the bounds for g21 problem.
%
% Returns:
% bounds: [lower, upper] for each variable

bounds = repmat([0, 10], 7, 1);

% Example usage
if ~exist('x', 'var')
    % Example solution (best known)
    x_opt = 0.1 * ones(7, 1);
    f_opt = g21_objective(x_opt);
    g_opt = g21_constraints(x_opt);
    
    fprintf('Objective value: %f\n', f_opt);
    fprintf('Constraints: ');
    disp(g_opt');
    fprintf('All constraints satisfied: %d\n', all(g_opt <= 0));
end